src/main/scala/io/circe/yaml/parser/NodeAlg.scala

```diff
@@ -0,0 +1,164 @@
+package io.circe.yaml.parser
+
+import cats.data.ValidatedNel
+import cats.instances.list._
+import cats.syntax.either._
+import cats.syntax.traverse._
+import io.circe.{Json, JsonNumber, JsonObject, ParsingFailure}
+import org.yaml.snakeyaml.constructor.SafeConstructor
+import org.yaml.snakeyaml.nodes._
+import scala.collection.JavaConverters._
+import scala.collection.immutable.Queue
+
+abstract class NodeAlg[T] {
+  def int(node: ScalarNode): T
+  def float(node: ScalarNode): T
+  def timestamp(node: ScalarNode): T
+  def bool(node: ScalarNode): T
+  def yNull(node: ScalarNode): T
+  def string(node: ScalarNode): T
+  def otherScalar(node: ScalarNode): T
+  
+  def sequence(node: SequenceNode): T = fromValues {
+    node.getValue.asScala.foldLeft(Queue.empty[T]) {
+      (accum, next) => accum enqueue any(next)
+    }
+  }
+  
+  def mapping(node: MappingNode): T = fromFields {
+    node.getValue.asScala.map {
+      nodeTuple => nodeTuple.getKeyNode match {
+        case keyNode: ScalarNode => keyNode.getValue -> any(nodeTuple.getValueNode)
+        case _ => throw ParsingFailure("Only string keys can be represented in JSON", null)
```

Comment:
I could just give it an actual exception here. I don't think you should change `ParsingFailure`.